### PR TITLE
First iteration of dict anti-corruption layer

### PR DESCRIPTION
**Pull request by Jim-Holmstroem**
Just a suggestion (could be filled in whit a lot more things), what do you think?

we could also start using https://toolz.readthedocs.io/en/latest/api.html

### CODE REVIEW

**Comment by Jim-Holmstroem** [comment]
just realized filter doesn't work since I haven't imported star, but the point is the point

**Comment by samedii** [comment]
I commented on this in slack but I'll repeat it here.

1. This function `dict.get(key, default=None)` exists in python's built-in `dict`
2. Would chaining over dictionaries be useful for us?
3. `Dict.map` is returning a `dict` rather than a `Dict`. Is this intended?
4. Setting and getting attributes could be nice
5. Overloading operators is cool but I don't think it will be used? What do you think? And what do we think of having unused methods?
6. Does toolz support all this that you propose or is it missing some feature?

File: workflow/functional/dict.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        '''Dict(a=3).a = 1
+        '''
+        return super().__setitem__(item, value)
+
+    def __add__(self, other):
+        '''Dict(a=3) + Dict(b=4)
+        '''
+        return Dict(self, **other)
+
+    def get(self, key, default=None):
```
Comment (samedii) [inline workflow/functional/dict.py:30]: I think this function should be removed as it already exists `dict`

File: workflow/functional/dict.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        if key in self:
+            return self[key]
+        else:
+            return default
+
+    def add(self, *args, **kwargs):
+        return self + Dict(*args, **kwargs)
+
+    def map(self, f):
+        return dict(starmap(f, self.items()))
```
Comment (samedii) [inline workflow/functional/dict.py:40]: I think this function should return a `Dict` so we can chain functions nicely

**Comment by Jim-Holmstroem** [comment]
1. Oops, it does, I'm a bit rusty, the reason I wrote it was because .get(.) wheren't used in the code on quite a few places
2. personally I think it's cleaner, but this was mostly designed as a discussion piece
3. No that is a bug
4. -
5. It can be nice sometimes not sure if this is the case, but I always miss having operators for things that makes sense (like adding 2 dictionaries)
6. yes toolz.dicttoolz support most of this but more on the form `operation(dict, *args)`

**Comment by samedii** [comment]
1. Do you mean we were doing checks for keys in dictionaries and then getting values or do you prefer `my_dict.get('some_key')` over my_dict['some_key']?
2. Cleaner than what? Do you think we are iterating over `dict` a lot now? If we are then I think this could be useful
5. I only think multiplication could be useful for something like mixup/mixmatch if we have `dict`s in our data pipeline. Otherwise, I don't see us using the same operator for every key
6. Okay I see. I'll have to look at some use cases

**Comment by Jim-Holmstroem** [comment]
Would also be possible to have the same type of thing for "Pipe" or something
```
thingy = Pipe(repeat)
    .then_map(fn)
    .then(chain.from_iterable)

thingy(input)
```

```
renderings = Pipe(id)
    .then_map(infer)
    .reduce(statistics)
    .then_map(plot)

renderings(input)
```

**Comment by samedii** [comment]
> Would also be possible to have the same type of thing for "Pipe" or something

Yes I like this. Let's look for a library that does something like this and has a decent user base if possible

**Comment by Jim-Holmstroem** [comment]
1. looked through the usage we have on this, and it was almost always something else when we checked for existence first.
in Rust they even have a "default value is retrieved from f() if needed" so that you don't have to spend cycles unless you need to
2. Cleaner in the sense that it's more functional then typical pythonic code, not that uncommon to have like for example dictionary mapping id to images and then you want to "map" the images without extra boilerplate

**Comment by Jim-Holmstroem** [comment]
either if we find something or create something new I can steal the usable parts of these things from Rust and pandas

**Comment by samedii** [comment]
1. Okay, let's try to keep this in mind and use `dict.get` more. I can't see any usage in coop-article-detection but I see some in ml-workflow. I can make a PR to fix this
2. I see. I think I have been using `pandas.DataFrame` for the use case that you are describing and it allows chaining functions too.

**Comment by Jim-Holmstroem** [comment]
Closing this one for now, we need to check for alternatives and figure out what we want

**Comment by samedii** [comment]
Okay, I think this brings up a good point though. We should look into this more and maybe come back to this code later

**Pull request closed without merging**